### ISSUE TITLE
Decide style re: "top level" objects

### ISSUE THREAD

**Issue by cjerdonek** [votinginfoproject/vip-specification#76]
I think it would be good if the style guide documented how to handle the "top-level" object collections, since this comes up a lot.

1622.2 [currently defines](http://grouper.ieee.org/groups/1622/groups/2/WorkingDocuments/V39/ElectionResultsV39.xsd) a named type for each such type, and then a top-level "Collection" element.  For example--

```xml
<xsd:complexType name="Candidate">
  <!-- stuff -->
</xsd:complexType>

<xsd:element name="CandidateCollection" minOccurs="0">
  <xsd:complexType>
    <xsd:sequence>
      <xsd:element name="Candidate" type="Candidate" minOccurs="0" maxOccurs="unbounded"/>
    </xsd:sequence>
  </xsd:complexType>
</xsd:element>
```

And VIP often "in-lines" the type definition with an anonymous complex type.  Also, VIP doesn't seem to wrap the sequence in a container node.  It just keeps everything flat at the top.  For example--

```xml
<xs:element name="Candidate">
  <xs:complexType>
    <!-- stuff -->
    <xs:attribute name="id" type="xs:ID" use="required"/>
  </xs:complexType>
</xs:element>
```

(Side note: shouldn't the element line have a "maxOccurs=unbounded"?)

Personally, I like 1622.2's approach.  I also prefer putting the sequences in a container node, which could still be done, even if VIP continues to use the anonymous type approach.

**Comment by cjerdonek** [comment]
@jungshadow Do you have any opinions on this?  I like 1622.2's "named type" approach because it breaks things up more, and the type definitions become portable (e.g. can be moved around without affecting the structure of the root "instance" object).  It would also make comparison with 1622.2 easier if the type definitions are separate from the elements.

**Comment by jdmgoogle** [comment]
I believe @pstenbjorn had some comments about this, so I'll leave it to him to hash out the details of the XMLisms.

**Comment by pstenbjorn** [comment]
@jdmgoogle and @cjerdonek there are two different xsd schema construction approaches that we are blending (Russian Doll and Venetian Blind).  Usually it best to chose a design method and use it throughout.  What @cjerdonek is recommending is using  Venetian Blind model (which is what 1622.2 uses).  Generally, Russian Doll is chosen when human readability of the schema and resulting XML is important and Venetian Blind is handy if there are multiple reusable elements.  

VIP historically has been Russian Doll and my comments have been that we should chose a design method that works for us.  In the merging of the ballot items into the VIP schema, this started to get ugly in that there where Venetian Blind techniques (e.g. extending named complex types) that created a hybrid schema. 

What we decide should be driven by the use case of the data.  Generally speaking our data are not series of related elements (while there is a growing degree of relationships in the data).  So I would be happy to begin the transition to a Venetian Blind design for elements that are reused within the schema.  

And to answer @cjerdonek specific question about the `maxOccurs` for each element. With the current Russian Doll schema structure, the parent element `vip_object` contains a `choice` container with an attribute of `maxOccurs="unbounded"` which means that the elements within the choice container can appear `unbounded` times unless the `maxOccurs=1`.

So in short, my suggestion is to chose a methodology, and a coding standard which allows us to violate its principles based upon specific use cases.

**Comment by cjerdonek** [comment]
@pstenbjorn So to restate your suggestion, are you saying that the guideline could be that type definitions be separated out (a la the Venetian Blind approach) if and only if they are instantiated in more one location in the schema, or are you saying if and only if the type definition is referenced via a `type="xs:IDREF"`?  One benefit of the latter approach is that it becomes clearer which elements are candidates for being referenced.

**Comment by pstenbjorn** [comment]
@cjerdonek what I was not very concisely trying to say is whatever approach we take we should be rule driven.  I think your rule (if an element is referred to by its id in another element) is a possible rule to apply.  Not that I am advocating against this rule, but if we applied it it would change most elements in the schema, which might be a good thing.

**Comment by cjerdonek** [comment]
Okay, thanks, @pstenbjorn.  Then I would like to propose the following as a starting point on this issue:

A type should be defined outside of the root element (a la Venetian Blind), if--

1. elements of that type are referenced by ID via `type="xs:IDREF"` (e.g. Candidate, PollingLocation, etc),
2. elements of that type occur in more than one location within the root element (e.g. InternationalizedText , SimpleAddressType, etc), or
3. the type is an "enum" (e.g. CandidatePreElectionStatus, TermType, etc).

(The particular wording can change, but this is the general idea.  Also, it's possible that I'm missing certain cases and that we may want to adjust this over time.)

In addition, we can specify how the type definitions should be ordered to remove that ambiguity.  VSSC seems to order type definitions something like: enums first, then complex types, and ordered alphabetically by name within each of those categories.  However, we can work that out separately.

Some benefits of the approach above are--

1. It will break up the main VipObject root element and make it less monolithic.
2. It will make our schema easier to compare and update with respect to VSSC 1622.2 (e.g. type definitions will be comparable one-to-one, and in some cases type definitions could potentially be cut-and-pasted from VSSC to VIP without change, and vice versa).
3. There will probably be less commit churn and smaller commit diffs (e.g. moving a node within the root element will not also involve moving the type definition, and vice versa).
4. This would also open up the possibility of putting type definitions in a separate file if we wanted to (e.g. to share them more easily), though I'm not advocating this.

**Comment by jungshadow** [comment]
@cjerdonek The pros sound good. Any cons? I might be able to see some issues from the processing- and readability-sides, but I'll let @pkoms, @nomadaisy, and @pstenbjorn advocate for themselves there.

**Comment by cjerdonek** [comment]
I'll let others weigh in, but the main cons I can think of are--

1. It will mean a fairly large change right now, which @pstenbjorn already pointed out.
2. You will probably need to "jump around" more when reading the spec (though on the flip side there will be less to scroll through if you are reading the root node).
3.  Also, [this page](http://www.oracle.com/technetwork/java/design-patterns-142138.html) has a table of pros and cons for the various patterns that @pstenbjorn mentioned.  The only con it has for Venetian Doll is, "Limits encapsulation by exposing types," though I'm not 100% sure what that means.

**Comment by jdmgoogle** [comment]
My vote is in favor of the Venetian Blind approach, although I agree that should be one of the final changes after we've figured out the actual objects themselves and the relationships between said objects.

I'm also putting together a document listing out the final changes we need to make in order to get this release candidate out the door. I'll share that document with everyone momentarily.

I'd also like to do another in-person/over-Hangout meeting with everyone where we can just go through the list of remaining items and knock them out that way.

**Comment by jungshadow** [comment]
@jdmgoogle The meeting sounds good. One goal I'd like to come out of this meeting (whether I'm there or not) is determining exactly where we draw the line with this 5.0 release. At this point, I'm hoping it encompasses very few spec-related issues and we move additional substantive changes to 5.1.

As for determining Venetian Blind or Russian Doll, we still need to hear from @nomadaisy or @pkoms or other Democracy Works folks about how this major structural change will affect the harvesting/processing of the data and the structure of the CSVs. I imagine it will be significant, but I also don't want to speak for them.

**Comment by cjerdonek** [comment]
@jungshadow Maybe @pstenbjorn would want to chime in on this, too, but I'm fairly certain this wouldn't affect the structure of the CSV's at all (e.g. since the type definitions would remain the same).  It also seems like this wouldn't change the format of the XML files.  I would view it more as a "refactoring" of the current schema (i.e. in practice, it should be equivalent in terms of the XML file format is spells out).  @pstenbjorn, would you agree?

**Comment by pstenbjorn** [comment]
@cjerdonek in general I agree that there is little correlation between the complexity of structuring CSV files and the top level style of the schema.  I think @jungshadow was concerned about the complexity of the XML generation process and how it might affect DW XML creation processes.

**Comment by nomadaisy** [comment]
Just an update on this - we are discussing internally about what this would require for processing and output in parallel to the 3.0 spec. We should have more information on Monday.

**Comment by nomadaisy** [comment]
Our team is in agreement to go ahead with changing over to the Venetian Blind approach. We've talked about the various processing implications, and they are in line with our anticipated work on switching over to the 5.0 spec once it's ready. One point I'd like to bring up is we'll be asking vendors to reconfigure their XML exports for many states if we switch over to a Venetian Blind structure.

**Comment by cjerdonek** [comment]
That's great, @nomadaisy.  Thanks a lot for the update.

Out of curiosity, do you know for sure that the XML exports will actually need to change in some way with a Venetian Blind XSD?  My current sense is that there is a chance the XML exports won't need to change at all in response, and that the change to Venetian Blind only affects how the spec is organized (and not the format of the resulting XML files).  Thanks.

**Comment by cjerdonek** [comment]
I created PR #107 for this.

**Comment by cjerdonek** [comment]
Closed by PR #107.
